Subclasses

chrisflesher opened the pull request:
@jfjlaros The subclasses I think are an improvement. I tested this branch with the `SocketInterface` and made some minor fixes to get it working.

chrisflesher commented on simple_rpc/simple_rpc.py:
```diff
@@ -199,11 +203,11 @@ def _get_methods(self):
     def is_open(self):
         """Query interface state."""
-        return bool(self.methods)
+        return len(self.methods) > 0
```
This does the same thing but it was a bit easier to read.

chrisflesher commented on simple_rpc/simple_rpc.py:
```diff
@@ -175,13 +175,17 @@ def close(self):
 
+        if autoconnect:
+            self.open()
+            self._close()
```
There needed to be some way for `SocketInterface` to call `get_methods`. This is kind of similar to how `autoconnect` is working but since `_get_methods` is private it cannot be specified manually?

chrisflesher commented on simple_rpc/simple_rpc.py:
```diff
@@ -184,3 +142,78 @@ def call_method(self, name, *args):
+        """
+        super().__init__(device, baudrate)
+
+        self._get_methods()
```
I called `_get_methods` directly here. At first I wrote this using autoconnect, open, and close but decided to do it this way since you'd have to use private method calls to do it correctly.

jfjlaros commented:
Yes, that looks good to me. I merged your changes in the `subclasses` branch.

jfjlaros commented:
I noticed that only calling `_get_methods` does not set the method attributes on the class. I just pushed an other proposal.

chrisflesher commented:
I tried running the latest and got the error below. Revised the proposal to get things working.

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/simple_rpc/simple_rpc.py", line 40, in _open
    self._connection.open()
  File "/usr/local/lib/python3.5/dist-packages/serial/urlhandler/protocol_socket.py", line 58, in open
    raise SerialException("Port is already open.")
serial.serialutil.SerialException: Port is already open.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.5/dist-packages/simple_rpc/simple_rpc.py", line 184, in __init__
    self.open()
  File "/usr/local/lib/python3.5/dist-packages/simple_rpc/simple_rpc.py", line 190, in _auto_open_wrapper
    result = f(self, *args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/simple_rpc/simple_rpc.py", line 212, in open
    super().open()
  File "/usr/local/lib/python3.5/dist-packages/simple_rpc/simple_rpc.py", line 104, in open
    self.methods = self._get_methods()
  File "/usr/local/lib/python3.5/dist-packages/simple_rpc/simple_rpc.py", line 189, in _auto_open_wrapper
    self._open()
  File "/usr/local/lib/python3.5/dist-packages/simple_rpc/simple_rpc.py", line 42, in _open
    raise IOError(error.strerror.split(':')[0])
AttributeError: 'NoneType' object has no attribute 'split'
>>> i = simple_rpc.SocketInterface('socket://192.168.1.50:10000')
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/simple_rpc/simple_rpc.py", line 40, in _open
    self._connection.open()
  File "/usr/local/lib/python3.5/dist-packages/serial/urlhandler/protocol_socket.py", line 58, in open
    raise SerialException("Port is already open.")
serial.serialutil.SerialException: Port is already open.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.5/dist-packages/simple_rpc/simple_rpc.py", line 184, in __init__
    self.open()
  File "/usr/local/lib/python3.5/dist-packages/simple_rpc/simple_rpc.py", line 190, in _auto_open_wrapper
    result = f(self, *args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/simple_rpc/simple_rpc.py", line 212, in open
    super().open()
  File "/usr/local/lib/python3.5/dist-packages/simple_rpc/simple_rpc.py", line 104, in open
    self.methods = self._get_methods()
  File "/usr/local/lib/python3.5/dist-packages/simple_rpc/simple_rpc.py", line 189, in _auto_open_wrapper
    self._open()
  File "/usr/local/lib/python3.5/dist-packages/simple_rpc/simple_rpc.py", line 42, in _open
    raise IOError(error.strerror.split(':')[0])
AttributeError: 'NoneType' object has no attribute 'split'
```

jfjlaros commented on simple_rpc/simple_rpc.py:
```diff
@@ -182,6 +182,7 @@ def __init__(self, device, baudrate=9600):
         """
         super().__init__(device, baudrate)
         self.open()
+        self._close()
```
I don't get it. The `open` function is wrapped by `_auto_open`, so a double close should not be needed, right?

chrisflesher replied:
Sorry this was a bit hard to see. I removed the `SocketInterface` open because I assumed it had some kind of recursive problem given the traceback message I got. I didn't look at it very closely though to see what the cause was.

jfjlaros replied:
Ah, I see. Both `open` and `_get_methods` are now wrapped and `open` calls `_get_methods`. Removing the method override for `_get_methods` should do the trick.

Can you test the latest version?

chrisflesher replied:
Sorry for the late reply, have been out on vacation the past few days. The latest version seems to work well, thanks!

jfjlaros replied:
I did not expect anything before the new year. Thanks for testing.

I added a little wrapper for backwards compatibility and to make the CLI work again. I will add some documentation later.

chrisflesher commented:
I tested this new wrapper with `python` and `simple_rpc` interfaces for both WiFi and serial and they seem to work.

jfjlaros commented:
Then I will proceed to merge the changes.
